Sources/ViewType.swift

```diff
@@ -0,0 +1,225 @@
+        case (.view(let lhs), .view(let rhs)):
+            return lhs == rhs
+        case (.layer(let lhs), .layer(let rhs)):
+            return lhs == rhs
+        default:
+            return false
+        }
+    }
+
+    public static func === (lhs: ViewType, rhs: ViewType) -> Bool {
+        switch (lhs, rhs) {
+        case (.view(let lhs), .view(let rhs)):
+            return lhs === rhs
+        case (.layer(let lhs), .layer(let rhs)):
+            return lhs === rhs
+        default:
+            return false
+        }
+    }
+}
+
+protocol Layout: class {
+    var superview: ViewType? { get }
+    var subviews: [ViewType] { get }
+    var frame: CGRect { get set}
+    var bounds: CGRect { get set}
+    var cornerRadius: CGFloat { get set}
+
+    func contains(_ view: ViewType) -> Bool
+    func convert(_ rect: CGRect, from view: ViewType?) -> CGRect
+}
+
+extension Layout {
+    func contains(_ view: ViewType) -> Bool {
+        if subviews.contains(view) {
+            return true
+        }
+        for subview in subviews {
+            if subview.layout.contains(view) {
+                return true
+            }
+        }
+        return false
+    }
+}
+
+final class UIViewLayout: Layout {
+    unowned let view: UIView
+
+    var nx_state: AnyHashable {
+        get {
+            view.nx_state
+        }
+        set {
+            view.nx_state = newValue
+        }
+    }
+
+    var superview: ViewType? {
+        guard let superview = view.superview else {
+            return nil
+        }
+        return .view(superview)
+    }
+
+    var subviews: [ViewType] {
+        return view.subviews.map { view in
+            .view(view)
+        }
+    }
+
+    var frame: CGRect {
+        get {
+            view.frame
+        }
+        set {
+            view.frame = newValue
+        }
+    }
+
+    var bounds: CGRect {
+        get {
+            view.bounds
+        }
+        set {
+            view.bounds = newValue
+        }
+    }
+
+    var cornerRadius: CGFloat {
+        get {
+            view.layer.cornerRadius
+        }
+        set {
+            view.layer.cornerRadius = newValue
+        }
+    }
+
+    func convert(_ rect: CGRect, from view: ViewType?) -> CGRect {
+        guard let view = view else {
+            return self.view.convert(rect, from: nil)
+        }
+        switch view {
+        case let .view(view):
+            return self.view.convert(rect, from: view)
+        case let .layer(layer):
+            return self.view.layer.convert(rect, from: layer)
+        }
+    }
+
+    func sizeToFit() {
+        view.sizeToFit()
+    }
+
+    func sizeThatFits(_ size: CGSize) -> CGSize {
+        view.sizeThatFits(size)
+    }
+
+    var safeAreaInsets: UIEdgeInsets {
+        if #available(iOS 11.0, *) {
+            return view.safeAreaInsets
+        }
+        return .zero
+    }
+
+    init(view: UIView) {
+        self.view = view
+    }
+}
+
+final class CALayerLayout: Layout {
```

Comment:
Naming is incorrect, this class is not Obj-C and it doesn't belong to CoreAnimation namespace.

Reply:
This naming assumes that this class is layout for CALayer. What is the name it should have by your opinion

Reply:
Just `ViewLayout` / `LayerLayout` would do just fine. I completely understand what's intention here, it's just mixing naming schemes what's unfortunate here. Since UIView & CALayer are Obj-C classes, UI & CA are class prefixes. And we don't use class prefixes in swift code, especially class prefixes of existing frameworks.